lib/src/git_backend.rs

```diff
@@ -606,6 +607,88 @@ fn to_no_gc_ref_update(id: &CommitId) -> gix::refs::transaction::RefEdit {
     }
 }
 
+fn to_ref_deletion(git_ref: gix::refs::Reference) -> gix::refs::transaction::RefEdit {
+    let expected = gix::refs::transaction::PreviousValue::ExistingMustMatch(git_ref.target);
+    gix::refs::transaction::RefEdit {
+        change: gix::refs::transaction::Change::Delete {
+            expected,
+            log: gix::refs::transaction::RefLog::AndReference,
+        },
+        name: git_ref.name,
+        deref: false,
+    }
+}
+
+/// Recreates `refs/jj/keep` refs for the `new_heads`, and removes the other
+/// unreachable and non-head refs.
+fn recreate_no_gc_refs(
+    git_repo: &gix::Repository,
+    new_heads: impl IntoIterator<Item = CommitId>,
+    keep_newer: SystemTime,
+) -> Result<(), BackendError> {
+    // Calculate diff between existing no-gc refs and new heads.
+    let new_heads: HashSet<CommitId> = new_heads.into_iter().collect();
+    let mut no_gc_refs_to_keep_count: usize = 0;
+    let mut no_gc_refs_to_delete: Vec<gix::refs::Reference> = Vec::new();
+    let git_references = git_repo
+        .references()
+        .map_err(|err| BackendError::Other(err.into()))?;
+    let no_gc_refs_iter = git_references
+        .prefixed(NO_GC_REF_NAMESPACE)
+        .map_err(|err| BackendError::Other(err.into()))?;
+    for git_ref in no_gc_refs_iter {
+        let git_ref = git_ref.map_err(BackendError::Other)?.detach();
+        let oid = git_ref.target.try_id().ok_or_else(|| {
+            let name = git_ref.name.as_bstr();
+            BackendError::Other(format!("Symbolic no-gc ref found: {name}").into())
+        })?;
+        let id = CommitId::from_bytes(oid.as_bytes());
+        let name_good = git_ref.name.as_bstr()[NO_GC_REF_NAMESPACE.len()..] == id.hex();
+        if new_heads.contains(&id) && name_good {
+            no_gc_refs_to_keep_count += 1;
+            continue;
+        }
+        // Check timestamp of loose ref, but this is still racy on re-import
+        // because:
+        // - existing packed ref won't be demoted to loose ref
+        // - existing loose ref won't be touched
+        //
+        // TODO: might be better to switch to a dummy merge, where new no-gc ref
+        // will always have a unique name. Doing that with the current
+        // ref-per-head strategy would increase the number of the no-gc refs.
+        // https://github.com/martinvonz/jj/pull/2659#issuecomment-1837057782
```

Comment:
In addition to that, this would delete excessive no-gc refs if the git object store is shared by multiple jj repositories (or multiple operation logs.) I don't think that's practical setup, but is worth mentioning.